inc/Parsing/Lexer/Lexer.php

```diff
@@ -149,6 +149,13 @@ public function parse($raw)
             if ($currentLength === $length) {
                 return false;
             }
+            // If we are closing a block and the last character consumed by our matched
+            //  string is a newline, put it back.  See the following for details:
+            //  https://github.com/dokuwiki/dokuwiki/issues/4054
+            if ($mode == "__exit" && substr($matched, -1) == "\n") {
```

Comment:
I am by no means an expert on the DW Lexer — but should this not be:
`if ($this->isModeEnd($mode) && subset($matched, -1) == "\n") {`

Or better yet:
`if ($this->isModeEnd($mode) && str_ends_with($matched, "\n")) {`

Reply:
AIUI, str_ends_with is PHP8, and DokuWiki still support PHP 7.

As for isModeEnd:  you could be right;  I too am not an expert on the lexer.  A quick look at the source shows that the code for that function is:  `return ($mode === "__exit");` so it seems that they will be functionally equivalent.  For a person unfamiliar with the lexer, `$mode` was right there... :)

Reply:
Recently, in the development version polyfills for str_ends_with() and likes are added e.g. https://github.com/dokuwiki/dokuwiki/blob/2c6bdf5eec8075a71fd937c32972491f139de86f/inc/compatibility.php#L116-L117 so it fine to use them already.

Reply:
No a polyfill has been added recently. So str_*_with functions can now be used.
See https://github.com/dokuwiki/dokuwiki/blob/2c6bdf5eec8075a71fd937c32972491f139de86f/inc/compatibility.php#L90-L121

Reply:
Well, it doesn't work on my system...  I'm running 2023-04-04a "Jack Jackrum", not a development snapshot.  I'm assuming that functionality came post-Jack?  (And I really don't want to set up a development snapshot at this time:  in fact, I'm just manually patching this stuff in the GitHub web UI...)  But if the polyfill capability *is* in the post-Jack snapshot feel free to make that change.

Reply:
Well the polyfill was indeed added post 2023-04-04a ;-) So you would need to add that to your DW as well (or use the bleeding edge GitHub master branch instead of the released version, or use PHP 8+).

Actually I'm not sure how to add commits to an unmerged PR that I don't own? Ah, figured it out :-)

Reply:
Believe it or not, I'm doing this on a production system -- after all, I did this because I *needed* this on the production system!  So no, major changes or development snapshots are not really an option.  But like I said:  feel free to add commits to the PR -- or simply make the changes outside of my PR if you prefer.

But to both of you while I have your attention:  if you me to take a pass at the unit tests I will try -- but I need someone to tell me how to create the expected state graph.  I can't do it by hand, and I'm sure no one else does it by hand either.  But I need help to find where to get that information.

Reply:
I feel out of my depth regarding the unit tests. I took a look at https://github.com/dokuwiki/dokuwiki/blob/master/_test/tests/inc/parser/lexer.test.php but saw no easy way to add tests for this PR. But maybe I'm looking in the wrong place?

Reply:
I'm not sure how to add a test for the *lexer*, but we *can* add tests for the parse handers for listblock, preformatted and tables.  Look above for the comment from @Klap-in with proper paths.  In that case, we can give it a chunk of markup source, and then compare our statically-added state graph against the graph generated by the running code.  I replied with proposed source that covers all of the updated transitions above in this issue.  I also took that same markup and replaced the comments with heading tags and put it on my production server with my PR applied.  You can see the result here:  https://www.fluidtechservices.com/dokuwiki/blocktest

It shows that the markup is being processed correctly.  As opposed to what happens when you paste it into the DokuWiki sandbox, which I've also done:  https://www.dokuwiki.org/sandbox:playground

You can see on my server each type of markup results in exactly the right type of HTML.  You can see on the sandbox it does *not*.

Also, I searched for some doc on unit testing and found this:  https://www.dokuwiki.org/devel:unittesting  I can tell you this, I do not have the environment for setting this up -- I'm *way* over my time budget for this.  But if someone can give me a straightforward way to get the state graph, I can use the existing unit tests as a pattern and create new ones.

Reply:
You are correct:  I just noticed the same thing.  It seems the lexer is going into the preformatted block and grabbing the text, but not outputting the text as HTML.

After spending the last half-hour looking into it, I checked the sandbox:  this bug is in the current lexer as well.  So I don't know where that text is going, or why, but it's not related to this patch.  That problem already exists...

Reply:
In general markup level makes sense for me. However, as this feels as a rather general change, probably adding tests for the lexer as well might make sense for how it should consume/not consume new lines between different parsermodes. Of course, these tests are less trivial.

On short notice, I have not much time available. I hope I can find some time in a week/2 weeks.

Everybody who can create unit tests is invited to add them, I think especially for markup these are quite doable :-)